src/passes/TraceCalls.cpp

```diff
@@ -0,0 +1,229 @@
+      trackerCallParams.insert(
+        trackerCallParams.begin(),
+        builder.makeLocalTee(resultLocal, realCall, resultType));
+
+      replaceCurrent(builder.makeBlock(
+        {builder.makeCall(
+           wrapperName, trackerCallParams, Type::BasicType::none),
+         builder.makeLocalGet(resultLocal, resultType)}));
+    } else
+      replaceCurrent(builder.makeBlock(
+        {realCall,
+         builder.makeCall(
+           wrapperName, trackerCallParams, Type::BasicType::none)}));
+  }
+
+  std::set<TracedFunction> tracedFunctions;
+};
+
+struct TraceCalls : public Pass {
+  // Adds calls to new imports.
+  bool addsEffects() override { return true; }
+
+  void run(Module* module) override {
+    auto functionsDefinitions = getPassOptions().getArgument(
+      "trace-calls",
+      "TraceCalls usage: wasm-opt "
+      "--trace-calls=FUNCTION_TO_TRACE[:TRACER_NAME][,RESULT1_TYPE]"
+      "[,PARAM1_TYPE[,PARAM2_TYPE[,...]]][;...]");
+
+    auto tracedFunctions = parseArgument(functionsDefinitions);
+
+    for (const auto& tracedFunction : tracedFunctions) {
+      addImport(module, tracedFunction);
+    }
+
+    AddTraceWrappers(std::move(tracedFunctions)).run(getPassRunner(), module);
+  }
+
+private:
+  std::set<TracedFunction> parseArgument(const std::string& arg) {
+    std::set<TracedFunction> tracedFunctions;
+
+    for (const auto& definition : String::Split(arg, ";")) {
+      auto parts = String::Split(definition, ",");
+      if (parts.size() == 0) {
+        // empty definition, ignore
+        continue;
+      }
+
+      std::string originName, traceName;
+      parseFunctionName(parts[0], originName, traceName);
+
+      std::vector<Type> paramsAndResults;
+      for (size_t i = 1; i < parts.size(); i++) {
+        paramsAndResults.push_back(strToType(parts[i]));
+      }
+
+      tracedFunctions.emplace(TracedFunction{
+        Name(originName), Name(traceName), paramsAndResults, Type::none});
+    }
+
+    return tracedFunctions;
+  }
+
+  void parseFunctionName(const std::string& str,
+                         std::string& originName,
+                         std::string& traceName) {
+    auto parts = String::Split(str, ":");
+    switch (parts.size()) {
+      case 1:
+        originName = parts[0];
+        traceName = "trace_" + originName;
+        break;
+      case 2:
+        originName = parts[0];
+        traceName = parts[1];
+        break;
+      default:
+        Fatal() << "Failed to parse function name ('" << str
+                << "'): expected format FUNCTION_TO_TRACE[:TRACER_NAME]";
+    }
+  }
+
+  void addImport(Module* wasm, const TracedFunction& f) {
+    ImportInfo info(*wasm);
+
+    if (!info.getImportedFunction(ENV, f.tracerName)) {
+      auto import =
+        Builder::makeFunction(f.tracerName, Signature(f.params, f.results), {});
+      import->module = ENV;
+      import->base = f.tracerName;
+      wasm->addFunction(std::move(import));
+    }
+  }
+};
+
+// declare passes
+
```

Comment:
```suggestion
```

Reply:
Looks like we have this comment in other passes when there are multiple passes in a file, but not when there is just one, like here.